src/timezones/types.jl

```diff
@@ -0,0 +1,212 @@
+    )
+    i = max(i, 1)
+
+    n = length(t)
+    while i <= n && t[i].utc_datetime <= latest
+        utc_dt = from_utc ? local_dt : local_dt - offset(t[i].zone)
+
+        if utc_dt >= t[i].utc_datetime && (i == n || utc_dt < t[i + 1].utc_datetime)
+            push!(possible, (utc_dt, t[i].zone))
+        end
+
+        i += 1
+    end
+
+    return possible
+end
+
+function ZonedDateTime(local_dt::DateTime, tz::VariableTimeZone, occurrence::Int=0; from_utc::Bool=false)
+    possible = possible_dates(local_dt, tz; from_utc=from_utc)
+
+    num = length(possible)
+    if num == 1
+        utc_dt, zone = possible[1]
+        return ZonedDateTime(utc_dt, tz, zone)
+    elseif num == 0
+        throw(NonExistentTimeError(local_dt, tz))
+    elseif occurrence > 0
+        utc_dt, zone = possible[occurrence]
+        return ZonedDateTime(utc_dt, tz, zone)
+    else
+        throw(AmbiguousTimeError(local_dt, tz))
+    end
+end
+
+function ZonedDateTime(local_dt::DateTime, tz::VariableTimeZone, is_dst::Bool; from_utc::Bool=false)
+    possible = possible_dates(local_dt, tz; from_utc=from_utc)
+
+    num = length(possible)
+    if num == 1
+        utc_dt, zone = possible[1]
+        return ZonedDateTime(utc_dt, tz, zone)
+    elseif num == 0
+        throw(NonExistentTimeError(local_dt, tz))
+    elseif num == 2
+        mask = [zone.dst_offset > Second(0) for (utc_dt, zone) in possible]
+
+        # Mask is expected to be unambiguous.
+        !($)(mask...) && throw(AmbiguousTimeError(local_dt, tz))
+
+        occurrence = is_dst ? findfirst(mask) : findfirst(!mask)
+        utc_dt, zone = possible[occurrence]
+        return ZonedDateTime(utc_dt, tz, zone)
+    else
+        throw(AmbiguousTimeError(local_dt, tz))
+    end
+end
+
+function ZonedDateTime(local_dt::DateTime, tz::FixedTimeZone; from_utc::Bool=false)
+    utc_dt = from_utc ? local_dt : local_dt - offset(tz)
+    return ZonedDateTime(utc_dt, tz, tz)
+end
+
+function ZonedDateTime(zdt::ZonedDateTime, tz::VariableTimeZone)
+    i = searchsortedlast(
+        tz.transitions, zdt.utc_datetime,
+        by=v -> typeof(v) == Transition ? v.utc_datetime : v,
+    )
+
+    if i == 0
+        throw(NonExistentTimeError(localtime(zdt), tz))
+    end
+
+    zone = tz.transitions[i].zone
+    return ZonedDateTime(zdt.utc_datetime, tz, zone)
+end
+
+function ZonedDateTime(zdt::ZonedDateTime, tz::FixedTimeZone)
+    return ZonedDateTime(zdt.utc_datetime, tz, tz)
+end
+
+function DateTime(parts::Union{Period,TimeZone}...)
+    periods = Period[]
+    timezone = Nullable{TimeZone}()
+    for part in parts
+        if isa(part, Period)
+            push!(periods, part)
+        elseif isnull(timezone)
+            timezone = Nullable{TimeZone}(part)
+        else
+            error("Multiple timezones found")
+        end
+    end
+
+    dt = DateTime(periods...)
+    return isnull(timezone) ? dt : ZonedDateTime(dt, get(timezone))
+end
+
+type AmbiguousTimeError <: Exception
```

Comment:
Good point. That reminds me I wanted to make the parent of `AmbiguousTimeError` and `NonExistentTimeError` be `TimeError`.